sbi/inference/snpe/snpe_base.py

```diff
@@ -376,6 +290,103 @@ def _train(
+        # Update tensorboard and summary dict.
+        self._summarize(
+            round_=self._round, x_o=None, theta_bank=theta, x_bank=x,
+        )
+
+        # Update description for progress bar.
+        if show_train_summary:
+            print(self._describe_round(self._round, self._summary))
+
+        return deepcopy(self._neural_net)
+
+    def build_posterior(
+        self,
+        density_estimator: Optional[TorchModule] = None,
+        rejection_sampling_parameters: Optional[Dict[str, Any]] = None,
+        sample_with_mcmc: bool = False,
+        mcmc_method: str = "slice_np",
+        mcmc_parameters: Optional[Dict[str, Any]] = None,
+        copy_state_from: Optional[DirectPosteriorType] = None,
+    ) -> DirectPosterior:
+        r"""
+        Build posterior from the neural density estimator.
+
+        For SNPE, the posterior distribution that is returned here implements the
+        following functionality over the raw neural density estimator:<br/>
+        - correct the calculation of the log probability such that it compensates for
+            the leakage.<br/>
+        - reject samples that lie outside of the prior bounds.<br/>
+        - alternatively, if leakage is very high (which can happen for multi-round
+            SNPE), sample from the posterior with MCMC.
+
+        Args:
+            density_estimator: The density estimator that the posterior is based on.
+                If `None`, use the latest neural density estimator that was trained.
+            rejection_sampling_parameters: Dictionary overriding the default parameters
+                for rejection sampling. The following parameters are supported:
+                `max_sampling_batch_size` to set the batch size for drawing new
+                samples from the candidate distribution, e.g., the posterior. Larger
+                batch size speeds up sampling.
+            sample_with_mcmc: Whether to sample with MCMC. MCMC can be used to deal
+                with high leakage.
+            mcmc_method: Method used for MCMC sampling, one of `slice_np`, `slice`,
+                `hmc`, `nuts`. Currently defaults to `slice_np` for a custom numpy
+                implementation of slice sampling; select `hmc`, `nuts` or `slice` for
+                Pyro-based sampling.
+            mcmc_parameters: Dictionary overriding the default parameters for MCMC.
+                The following parameters are supported: `thin` to set the thinning
+                factor for the chain, `warmup_steps` to set the initial number of
+                samples to discard, `num_chains` for the number of chains,
+                `init_strategy` for the initialisation strategy for chains; `prior` will
+                draw init locations from prior, whereas `sir` will use
+                Sequential-Importance-Resampling using `init_strategy_num_candidates`
+                to find init locations.
+            copy_state_from: A previous posterior object from which the
```

Comment:
This sounds really confusing. I may need a call to understand the use case. Is there a tutorial / example with one?